certora/specs/SocialRecoveryModule.spec

```diff
@@ -1,26 +1,208 @@
+    requireGuardiansLinkedListIntegrity(guardian);
+
+    // The guardian should not be values such as zero, sentinel, or the Safe contract itself.
+    require guardian != 0;
+    require guardian != SENTINEL();
+    require guardian != safeContract;
+
+    // The guardian should not be an owner of the Safe contract at the time of addition.
+    require !safeContract.isOwner(guardian);
+    // The guardian should not be already added as a guardian.
+    require !currentContract.isGuardian(safeContract, guardian);
+
+    // The threshold must be greater than 0 and less than or equal to the total number of guardians.
+    require threshold > 0 && to_mathint(threshold) <= guardianStorageContract.entries[safeContract].count + 1;
+
+    // Safe contract should be the sender of the transaction.
+    require e.msg.sender == safeContract;
+    currentContract.addGuardianWithThreshold@withrevert(e, safeContract, guardian, threshold);
+    bool isReverted = lastReverted;
+
+    assert !isReverted &&
+        currentContract.isGuardian(safeContract, guardian) &&
+        currentGuardiansCount + 1 == to_mathint(guardianStorageContract.entries[safeContract].count);
+}
+
+// This integrity rule verifies the possibilites in which the addition of a new guardian can revert.
+rule addGuardianRevertPossibilities(env e, address guardian, uint256 threshold) {
+    bool isGuardian = currentContract.isGuardian(safeContract, guardian);
+
+    currentContract.addGuardianWithThreshold@withrevert(e, safeContract, guardian, threshold);
+    bool isReverted = lastReverted;
+
+    assert isReverted =>
+        isGuardian ||
+        e.msg.sender != safeContract ||
+        e.msg.value != 0 ||
+        guardian == 0 ||
+        guardian == SENTINEL() ||
+        guardian == safeContract ||
+        safeContract.isOwner(guardian) ||
+        threshold == 0 ||
+        threshold >= guardianStorageContract.entries[safeContract].count ||
+        guardianStorageContract.entries[safeContract].count == max_uint256 ||
+        !safeContract.isModuleEnabled(currentContract);
+}
+
+// This integrity rule verifies that if the revokeGuardianWithThreshold(...) executes, then ensure that:
+// - the Social Recovery Module is enabled
+// - the caller to the Module has to be the Safe Contract
+// - the guardian is revoked from the guardian list
+// - the linked list integrity remains,
+// - and no other account (guardian or not) is affected.
+rule revokeGuardiansWorksAsExpected(env e, address guardian, address prevGuardian, uint256 threshold, address otherAccount) {
+    address nextGuardian = guardianStorageContract.entries[safeContract].guardians[guardian];
+
+    bool otherAccountIsGuardian = currentContract.isGuardian(safeContract, otherAccount);
+
+    uint256 currentGuardiansCount = guardianStorageContract.entries[safeContract].count;
+    require currentGuardiansCount > 0;
+
+    requireGuardiansLinkedListIntegrity(guardian);
```

Comment:
uber-style-nit: Same as before, would be nice to be consistent in how we "present" the rules to make them easier to digest. I would suggest putting the assumptions all at the top.

Reply:
Done 👍🏾